tests/libvmi/vmi.go

```diff
@@ -0,0 +1,132 @@
+/*
+ * This file is part of the KubeVirt project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Copyright 2020 Red Hat, Inc.
+ *
+ */
+
+package libvmi
+
+import (
+	k8sv1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/api/resource"
+	k8smetav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/rand"
+
+	kvirtv1 "kubevirt.io/client-go/api/v1"
+)
+
+type Option func(vmiOptions)
+
+const DefaultVmiName = "testvmi"
+
+type action func(vmi *kvirtv1.VirtualMachineInstance)
+type vmiOptions map[int]action
+
+const defaultTestGracePeriod int64 = 0
+
+const (
+	// Options Identifiers
+	optNone = iota
+
+	optContainerImageName
+	optEnableRng
+	optResourceMemory
+	optInterface
+	optNetwork
+	optTerminationGracePeriod
+
+	optCloudInitNoCloudUserData
+	optCloudInitNoCloudUserDataB64
+	optCloudInitNoCloudNetworkData
+	optCloudInitNoCloudNetworkDataB64
+)
+
+var defaultVmiOptions = vmiOptions{
+	optInterface: func(vmi *kvirtv1.VirtualMachineInstance) {
+		vmi.Spec.Domain.Devices.Interfaces = append(
+			vmi.Spec.Domain.Devices.Interfaces, primaryInterfaceDevice(),
+		)
+	},
+	optNetwork: func(vmi *kvirtv1.VirtualMachineInstance) {
+		vmi.Spec.Networks = append(vmi.Spec.Networks, *kvirtv1.DefaultPodNetwork())
+	},
+	optTerminationGracePeriod: func(vmi *kvirtv1.VirtualMachineInstance) {
+		t := defaultTestGracePeriod
+		vmi.Spec.TerminationGracePeriodSeconds = &t
+	},
+	optResourceMemory: func(vmi *kvirtv1.VirtualMachineInstance) {
+		vmi.Spec.Domain.Resources.Requests = k8sv1.ResourceList{
+			k8sv1.ResourceMemory: resource.MustParse("8192Ki"),
+		}
+	},
+}
+
+// WithRng adds `rng` to the the vmi devices.
+func WithRng() Option {
+	return func(o vmiOptions) {
+		o[optEnableRng] = func(vmi *kvirtv1.VirtualMachineInstance) {
+			vmi.Spec.Domain.Devices.Rng = &kvirtv1.Rng{}
+		}
+	}
+}
+
+// WithResourceMemory specifies the vmi memory resource.
+func WithResourceMemory(value string) Option {
+	return func(o vmiOptions) {
+		o[optResourceMemory] = func(vmi *kvirtv1.VirtualMachineInstance) {
+			vmi.Spec.Domain.Resources.Requests = k8sv1.ResourceList{
+				k8sv1.ResourceMemory: resource.MustParse(value),
+			}
+		}
+	}
+}
+
+// New instantiates a new VMI configuration,
+// building its properties based on the specified With* options.
+func New(namespace, name string, opts ...Option) *kvirtv1.VirtualMachineInstance {
```

Comment:
It will look utterly weird and non understood.
You could argue that the name should come in randomized or not already, done by a helper the user can call.